### PR TITLE
Reddit search

### PR DESCRIPTION
**Pull request by AhmadAbdelaziz8**
## Description

built a comprehensive MCP (Model Context Protocol) server that provides tools capable of searching Reddit to enrich LLMs with contextual information from Reddit communities. This MCP server exposes four main tools: subreddit discovery, post searching with semantic ranking, comment retrieval, and similar post discovery using semantic matching.

Due to time constraints, ÷ focused on core Reddit search functionality rather than adding more writing tools, and the implementation would benefit from more rigorous testing with Claude and other LLM platforms.

## Type of change

- [x] New MCP feature (non-breaking change which adds functionality)

## How has this been tested?

The Reddit MCP server has been tested with:

- Basic functionality verification using direct Python execution
- Docker containerization testing
- Integration testing with Cursor IDE via MCP configuration
- Manual validation of all four core tools with MCP inspector:
  - `reddit_find_subreddits`: Finding relevant communities
  - `reddit_search_posts`: Searching posts with semantic ranking
  - `reddit_get_post_comments`: Retrieving top comments
  - `reddit_find_similar_posts`: Semantic similarity matching

**Note**: Due to time constraints, more rigorous testing with Claude and comprehensive edge case validation is still needed.

## Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings

2  tool

Tools
- [x] New and existing tests pass locally with my changes

### CODE REVIEW

**Comment by udaykakade25** [comment]
Hey @AhmadAbdelaziz8 , Thanks for the PR
Make sure your MCP server follows infra similar to other MCPs
klavisai_mcp_infra

**Comment by AhmadAbdelaziz8** [comment]
> Hey @AhmadAbdelaziz8 , Thanks for the PR Make sure your MCP server follows infra similar to other MCPs klavisai_mcp_infra

thanks you for making this note, I just refactored the Folder structure 

{BE31E181-B63E-48A8-A6DF-A2C807B1C6EE}

File: mcp_servers/reddit_search/uv.lock
Comment (Mayank-MSJ-Singh) [inline mcp_servers/reddit_search/uv.lock:1]: I think We can Remove this file since we have dockers already

File: mcp_servers/reddit_search/Dockerfile
Comment (Mayank-MSJ-Singh) [inline mcp_servers/reddit_search/Dockerfile:1]: this also

File: mcp_servers/reddit_search/.python-version
Comment (Mayank-MSJ-Singh) [inline mcp_servers/reddit_search/.python-version:1]: this one also

File: mcp_servers/reddit_search/tools/shared_utils.py [hunk 1]
```diff
@@ -0,0 +1,219 @@
+import os 
+import logging
+import requests
```
Comment (Mayank-MSJ-Singh) [inline mcp_servers/reddit_search/tools/shared_utils.py:3]: We shouldn’t use `requests` because it defeats async’s purpose. Instead, use `httpx` or `aiohttp`.
Here’s a good video explaining it: https://www.youtube.com/watch?v=OPyoXx0yA0I
I recommend `httpx` since MCP package needs it already, so no extra install like `aiohttp`.

**Comment by Mayank-MSJ-Singh** [comment]
@AhmadAbdelaziz8 Your code is good but there are some problems with it, it's less flexible, less scalable, 
I would like you to go through these MCP Servers (specially see there server.py, base.py and also tools),

https://github.com/Klavis-AI/klavis/tree/main/mcp_servers/hubspot
https://github.com/Klavis-AI/klavis/tree/main/mcp_servers/brave_search

If you need help, join our Discord: https://discord.gg/nJcKMQzkcW

File: mcp_servers/reddit_search/pyproject.toml
[the remark this answers is not in the document]
Comment (AhmadAbdelaziz8) [reply mcp_servers/reddit_search/pyproject.toml:1]: Thanks for the feedback, now they're removed and docker is kepy as the single source of truth

**Comment by AhmadAbdelaziz8** [comment]
Hey, I've completely refactored the code accordingly. 

- Migrated to an asynchronous httpx concurrent API calls.

- Replaced the FastMCP framework with a native low-level server implementation, to add dual transport support (SSE & StreamableHTTP)

- Removed redundant version and lock files, streamlining the project to a Docker-only configuration to reduce maintenance overhead.

- Tested the app once again in MCP client ( cursor in this case ), and it was quite useful for me

{135166DA-D954-41BA-9226-5DBD9CA53B29}

**Review by udaykakade25 — COMMENTED** [review]
LGTM!
Left some comments!
Let me know when you refactor, ready to merge!

**Comment by AhmadAbdelaziz8** [comment]
hey, I've submitted the following tools: 

reddit_create_post - Create new text posts
reddit_create_comment - Post comments
reddit_upvote - Upvote posts/comments

in conjunction with the current reading tools, they're having a synergistic effect on each other! 

here's a showcase of it's usage, this post, was powered by the Reddit mcp server from A  to Z, it founded the subreddit, write the post, comment and upvoting post, just by one prompt! 

here's the post: 
https://www.reddit.com/r/MachineLearningAndAI/comments/1nqc6gh/introducing_reddit_search_klavis_ais_opensource/

**Comment by zihaolin96** [comment]
only one QQ - I heard Reddit has strict rate limit on API, any thought on handle this?
